File: javascript/src/method/toMap.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"

import {CollectionConstants}                             from "../CollectionConstants"
import {__associativeValues, __associativeValuesByArray} from "./_tables utility"
import {isArray}                                         from "./isArray"
import {isArrayByStructure}                              from "./isArrayByStructure"
import {isCollectionHolder}                              from "./isCollectionHolder"
import {isCollectionHolderByStructure}                   from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}                    from "./isMinimalistCollectionHolder"

//#region -------------------- Facade method --------------------

/**
 * Convert the {@link collection} to an {@link ReadonlyMap map}
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array}) to convert
 * @extensionFunction
 */
export function toMap<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>,): ReadonlyMap<number, T> {
    if (collection == null)
        return CollectionConstants.EMPTY_MAP
    if (isCollectionHolder(collection,))
        return toMapByCollectionHolder(collection,)
    if (isArray(collection,))
        return toMapByArray(collection,)
    if (isMinimalistCollectionHolder(collection,))
        return toMapByMinimalistCollectionHolder(collection,)

    if (isCollectionHolderByStructure<T>(collection,))
        return toMapByCollectionHolder(collection,)
    if (isArrayByStructure<T>(collection,))
        return toMapByArray(collection,)
    return toMapByMinimalistCollectionHolder(collection,)
}


/**
 * Convert the {@link collection} to an {@link ReadonlyMap map}
 *
 * @param collection The {@link Nullable nullable} {@link MinimalistCollectionHolder collection} to convert
 * @extensionFunction
 */
export function toMapByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>,): ReadonlyMap<number, T> {
    if (collection == null)
        return CollectionConstants.EMPTY_MAP

    const size = collection.size
    if (size == 0)
        return CollectionConstants.EMPTY_MAP
    return __newMap(collection, size,)
}

/**
 * Convert the {@link collection} to an {@link ReadonlyMap map}
 *
 * @param collection The {@link Nullable nullable} {@link CollectionHolder collection} to convert
 * @extensionFunction
 */
export function toMapByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>,): ReadonlyMap<number, T> {
    if (collection == null)
        return CollectionConstants.EMPTY_MAP
    if (collection.isEmpty)
        return CollectionConstants.EMPTY_MAP
    return __newMap(collection, collection.size,)
}

/**
 * Convert the {@link collection} to an {@link ReadonlyMap map}
 *
 * @param collection The {@link Nullable nullable} {@link ReadonlyArray collection} to convert
 * @extensionFunction
 */
export function toMapByArray<const T, >(collection: Nullable<readonly T[]>,): ReadonlyMap<number, T> {
    if (collection == null)
        return CollectionConstants.EMPTY_MAP

    const size = collection.length
    if (size == 0)
        return CollectionConstants.EMPTY_MAP
    return __newMapByArray(collection, size,)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop method --------------------

function __newMap<const T, >(collection: MinimalistCollectionHolder<T>, size: number,) {
    return Object.freeze(new Map(__associativeValues(collection, size,),),)
}

function __newMapByArray<const T, >(collection: readonly T[], size: number,) {
    return Object.freeze(new Map(__associativeValuesByArray(collection, size,),),)
}

//#endregion -------------------- Loop method --------------------

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/ToMap.java
package joookiwi.collection.java.method;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import joookiwi.collection.java.extended.map.ArrayAsImmutableMap;
import joookiwi.collection.java.extended.map.EmptyMap;
import joookiwi.collection.java.extended.map.ImmutableMap;
import joookiwi.collection.java.extended.set.ImmutableSortedSet;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_NEW_1;

@NotNullByDefault
public final class ToMap
        extends UtilityWithTable {

    @Contract(ALWAYS_FAIL_0)
    private ToMap() {
        throw new ImpossibleConstructionException("The utility class “ToMap” cannot be constructed.", ToMap.class);
    }

    //#region -------------------- Facade methods --------------------

    /// Convert the `collection` to an [ImmutableMap]
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> ImmutableMap<Integer, T> toMap(final @Nullable MinimalistCollectionHolder<? extends T> collection) {
        if (collection == null)
            return EmptyMap.getInstance();

        final var size = collection.size();
        if (size == 0)
            return EmptyMap.getInstance();
        return new ArrayAsImmutableMap<>(_associativeValues(collection, size));
    }

    /// Convert the `collection` to an [ImmutableMap]
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> ImmutableMap<Integer, T> toMap(final @Nullable CollectionHolder<? extends T> collection) {
        if (collection == null)
            return EmptyMap.getInstance();
        if (collection.isEmpty())
            return EmptyMap.getInstance();
        return new ArrayAsImmutableMap<>(_associativeValues(collection, collection.size()));
    }

    /// Convert the `collection` to an [ImmutableMap]
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> ImmutableMap<Integer, T> toMap(final T @Nullable @Unmodifiable [] collection) {
        if (collection == null)
            return EmptyMap.getInstance();

        final var size = collection.length;
        if (size == 0)
            return EmptyMap.getInstance();
        return new ArrayAsImmutableMap<>(_associativeValues(collection, size));
    }

    //#endregion -------------------- Facade methods --------------------

}
